### ISSUE TITLE
document by which figs and tables get added to html, facilitate posting to github pages.

### ISSUE THREAD

**Issue by iantaylor-NOAA** [r4ss/r4ss#544]
## Background
GitHub Pages allows anyone with a github account to post any files to the web, including the r4ss html. See https://r4ss.github.io/r4ss/BigSkate/ for example and https://pages.github.com/ for info on the github implementation.

Here's an attempt to document how to do that.

## Current framework
**1. Move existing plots from older models to avoid overwriting** https://github.com/r4ss/r4ss/blob/main/R/SS_plots.R#L408-L437
- `SS_plots()` reads any csv files that start with `plotInfoTable` within the specified `plotdir` (which = `plots` by default).
- If a value in the `StartTime` column of any of those csv files does not exactly equal the start time of the model provided as `replist` to `SS_plots()` then the `plots` directory is renamed to something like `plots_StartTime_Tue_Jun_22_17.34.28_2021` where the value appended in first StartTime value in the first csv file that got read.

**2. Get info on new plots**
- Each function called by `SS_plots()` (e.g. `SSplotBiology()`)returns invisibly a table with details on the plots that got created by that function, e.g. https://github.com/r4ss/r4ss/blob/main/R/SS_plots.R#L497-L505.
- The info provided by each function is `file` (the filename of the png file) and `caption` (which can include html tags and be as complex or as simple as you wish.
- A third variable `category` is added before passing that table back to `SS_output()`, e.g. `bio` for the plots created by `SSplotBiology()`: https://github.com/r4ss/r4ss/blob/main/R/SSplotBiology.R#L1668-L1670. 
- The info for all plots is compiled within `SS_plots()` which then adds two new variables: `png_time` (current time value for `Sys.time()` applied to all plots just made) and `StartTime` (starting time of the model on which the plots were based). The table with the five variables (`file`, `caption`, `category`, `StartTime`, and `png_time` is written to a csv file with name based on the current time (`paste0("plotInfoTable_", format(png_time, "%d-%m-%Y_%H.%M.%OS4"), ".csv")`) where the `%OS4` was just added to use decimal seconds out to 4 decimal places to avoid cases where two files were written within the same second by multiple calls to `SS_plots()`.
   
**3. Create html files**
- The `SS_html` function gets called at the end of `SS_plots() and creates a new tab for each each unique value in `category` in whatever order they appear in any csv file located within the `plotdir` directory (which get read in alphabetical order). If the StartTime values don't match it will stop with a warning which you can override via `multimodel = TRUE`.
- A file `_SS_output.html` gets created with info on the model start time, the SS version, the r4ss installation, and any note provided by the `filenotes` argument to `SS_plot()` (which I only used for the first time yesterday). This file is associated with the category "Home" and I don't know what would happen if any other information were assigned to that category in one of the plotInfoTable csv files.
- A new html file for each category is created with a name set by `paste("_SS_output_", category, ".html", sep = "")` which includes information on all the plots or tables associated with that category in whatever order they appear in the csv files (which are read in alphabetical order). The tabbed output at the top of each html file includes references to all the other files.
- If there are duplicated file names in the csv files (e.g. from multiple calls to `SS_plots()` with different inputs), the one with the more recent `png_time` is used.
- The `_SS_output.html` gets automatically opened in the default browser (unless you run `SS_html(..., openfile = FALSE)` outside of `SS_plots()`.

## adding extra figures to the html view
- Multiple calls to `SS_plots() with different arguments are fine but if you add more panels to a given plot, you might have a lingering "page3" file that when you only want when all the files now fit on pages 1 and 2.
- Any other plots or tables created outside of r4ss that are described in a csv file with matching format should be automatically combined with the ones created by r4ss and if you run `SS_html(..., multimodel = TRUE)` then the StartTime should be ignored.
- Making a csv filename that is alphabetically earlier than the ones created by r4ss should put the extra plots above the other r4ss plots within that category.
- Adding any value for `category` other than the existing defaults should get you a new tab which will probably be in the order it appears among the alphabetically read csv files.

## getting the plots onto github
_to be added_

## future directions to make all this easier within r4ss
_to be added_

**Comment by k-doering-NOAA** [comment]
@iantaylor-NOAA , was the goal with this issue to create documentation to inform users on how to post r4ss output to github pages? 

If so, maybe adding to the vignette would help facilitate this?

The [github pages documentation](https://docs.github.com/en/pages) could help facilitate this, I think.

**Comment by iantaylor-NOAA** [comment]
@k-doering-NOAA, yes, I think a vignette makes sense. I had a typo in the issue name which made it ambiguous.
I also think that there may be code changes that could make the process easier, but initially it's just to document the process as started in the first comment on the issue.

**Comment by k-doering-NOAA** [comment]
@iantaylor-NOAA I got started on this issue. Thinking more about where this documentation belongs, I thought most of it should go right into the `SS_plots()` r oxygen documentation, in a `@details` section so it is easier for users to find. I added this in a9c374b. 

We could demo how users could add their own custom plots to the html output in a vignette - I added some ideas of how to do this in a4a13aa. I'm not sure it is worth the effort to do this? And if it is, if this should go right as a new section in the intro vignette or in a separate one?

@kellijohnson-NOAA, I know you have a lot of thoughts on documentation, so it would be great to get your take on this.

Finally, it looks like there isn't any advice about how to improve r4ss to make this easier for users to implement this workflow. The one that comes to mind for me is to implement a way to allow users to directly pass in their extra plots and information to `SS_plots()` so that the approach is less of a hack. Perhaps a separate issue should be opened for this, if deemed important enough.

**Comment by iantaylor-NOAA** [comment]
Thanks for getting started on this @k-doering-NOAA. I think the vignette structure looks good to me. I think the draft Pull Request is on the right track.

I don't remember what my ideas were for changes to r4ss, but one thing might be to either avoid or help users navigate this issue with the leading underscores in the html filenames:

from https://github.com/nwfsc-assess/lingcod/issues/95#issuecomment-867969456:
> The recent change in r4ss to put underscores at the front of the html files (e.g. `_SS_output_Bio.html`) works well for seeing those files at the top of a directory you open locally, but apparently causes problems for the Jekyll software that processes github pages. The work-around (discussed at https://github.blog/2009-12-29-bypassing-jekyll-on-github-pages/) is to add a `.nojekyll` file, which I did in [d4fed00](https://github.com/nwfsc-assess/lingcod/commit/d4fed004ce48c00ff2cde375078139efaf911d8a). That got the r4ss viewers working, but makes the index.md page no longer visible so I added an uglier index.html which got it working. We can refine this in the future but the links shouldn't change.

**Comment by okenk** [comment]
The repo for my assessment contained all of the report files from tons of model runs, so was quite large. As a result, the above workflow did not work, and the [github actions server ran out of space while checking out the repo](https://github.com/pfmc-assessments/canary_2023/actions/runs/5624478222/job/15241381600). 

To solve, this, I created a new branch of my repo, `reduced_size`, updated the .gitignore for this branch to ignore the entire directory with model runs, and untracked the model directory, using `git rm`. Then, I followed the above advice and moved the r4ss plots into a docs folder and created an index.html file. Finally, under settings, I changed the branch that github pages builds and deploys from. It no longer tries to check out the main branch, and everything works great.

In theory updating the `reduced_size` branch could be automated with github actions, but since github actions was having trouble checking out the main branch, I'm not actually sure if it's actually possible.
